liblangutil/Scanner.cpp

```diff
@@ -789,13 +802,22 @@ Token Scanner::scanString()
 				return setError(ScannerError::IllegalEscapeSequence);
 		}
 		else
+		{
+			// Report error on non-printable characters in string literals, however
+			// allow anything for unicode string literals, because their validity will
+			// be verified later (in the syntax checker).
+			//
+			// NOTE: remove dependency on locale and thus not use isprint() here
+			if (!_isUnicode && (static_cast<unsigned>(c) <= 0x1f || static_cast<unsigned>(c) > 0x7f))
+				return setError(ScannerError::IllegalCharacterInString);
 			addLiteralChar(c);
+		}
 	}
 	if (m_char != quote)
 		return setError(ScannerError::IllegalStringEndQuote);
 	literal.complete();
 	advance();  // consume quote
-	return Token::StringLiteral;
+	return _isUnicode ? Token::UnicodeStringLiteral : Token::StringLiteral;
```

Comment:
Do we still need this distinction?

Reply:
Yes because with your proposed rules it is still possible to craft non-Unicode compliant "ascii strings" using escape characters.

Reply:
You mean non-utf8-compliant? But that would be checked for both kinds of strings in the same way, wouldn't it?